src/runtime_src/core/include/experimental/xrt_queue.h

```diff
@@ -0,0 +1,297 @@
+      : m_content(new task_holder<Callable>(std::forward<Callable>(c)))
+    {}
+
+    task&
+    operator=(task&& rhs)
+    {
+      m_content = std::move(rhs.m_content);
+      return *this;
+    }
+
+    operator bool() const
+    {
+      return m_content!=nullptr;
+    }
+
+    void
+    execute()
+    {
+      m_content->execute();
+    }
+  };  // class queue::task
+
+public:
+
+  /**
+   * class event - type erased std::shared_future
+   *
+   * Wraps typed future return value returned when
+   * enqueueing an operation.
+   *
+   * Returned futures implicitly convert to an event and as such
+   * enqueuue return values can be stored in an event container if
+   * necessary.
+   *
+   * The event object is not needed where the typed future can be
+   * used.
+   */
+  class event
+  {
+    struct event_iholder
+    {
+      virtual ~event_iholder() {};
+      virtual void wait() const = 0;
+    };
+
+    // Wrap tyoed future
+    template <typename ValueType>
+    struct event_holder : public event_iholder
+    {
+      std::shared_future<ValueType> m_held;
+
+      event_holder(std::shared_future<ValueType>&& e)
+        : m_held(std::move(e))
+      {}
+
+      void wait() const
+      {
+        m_held.wait();
+      }
+    };
+
+    std::shared_ptr<event_iholder> m_content;
+
+  public:
+    event() : m_content(nullptr)
+    {}
+
+    event(event&& rhs) : m_content(std::move(rhs.m_content))
+    {}
+
+    event(const event& rhs) : m_content(rhs.m_content)
+    {}
+
+    // event() - event constructor for std::shared_future
+    //
+    // @e : shared_future to type erase
+    template <typename ValueType>
+    event(std::shared_future<ValueType> e)
+      : m_content(new event_holder<ValueType>(std::move(e)))
+    {}
+
+    event&
+    operator=(event&& rhs)
+    {
+      m_content = std::move(rhs.m_content);
+      return *this;
+    }
+
+    operator bool() const
+    {
+      return m_content!=nullptr;
+    }
+
+    void
+    wait() const
+    {
+      if (m_content)
+        m_content->wait();
+    }
+  };
+
+private:
+  // Add task to queue
+  void
+  add_task(task&& ev);
+
+public:
+  /**
+   * queue() - Constructor for queue object
+   *
+   * The queue is constructed with one consumer thread.
+   */
+  queue();
+
+  /**
+   * enqueue() - Enqueue a callable
+   *
+   * @param c
+   *   Callable function, typically a lambda
+   * @return
+   *   Future result of the function (std::future)
+   *
+   * A callable is an argument less lambda function.  The function is
+   * executed asynchronously by the queue consumer (worker thread)
+   * once all previous enqueued operations have completed.
+   *
+   * Upon completion the returned future becomes valid and will
+   * contain the return value of executing the lambda.
+   */
+  template <typename Callable>
+  auto
+  enqueue(Callable&& c)
+  {
+    using value_type = decltype(c());
+    std::packaged_task<value_type()> task{[cc = std::move(c)] { return cc(); }};
```

Comment:
It took me some time to understand why this was correct.
In this unevaluated context `value_type()` here is a type declaration, not a value construction. :-)